ai_tools/utils/document_processor.py

```
@@ -265,18 +376,174 @@ def _translate_batch_with_masking(self, batch: List[str], translate_fn,
-                except Exception:
-                    final.append(original)
+                raise ValueError(
+                    f"Unsafe placeholder response for translation unit {i}: "
+                    f"{validation.reason}"
+                )
 
         return final, validation_failures
 
+    @staticmethod
+    def _invoke_translate_func(translate_fn, source: List[str], target_lang: str,
+                               document_context: str = None):
```

Ruff (0.16.1)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)